honor-first-shutdown

ejb42 opened the pull request:
This pull request is will make the first shutdown request be the only shutdown request honored. The scenario I specifically ran into was a service configured as FailureAction=reboot. During a normal poweroff shutdown this service failed and changed the shutdown to a reboot type. This happened since systemd honored the last shutdown request received.
The two test files can be used to both verify the problem exists and verify the problem is fixed. This issue was seen on a fedora 31 machine using systemd v243.
Steps for the test files
- Install the service file and enable
- Install the shell script in /home (or any location - just remember to change the service file accordingly)
reboot your systemd image so the new service is started normally
perform a poweroff to see it change to a reboot
Install this fix and perform the same steps and see the poweroff complete normally

keszybz commented:
This solution seems pretty intrusive. The original problem was that a FailureAction= from a failing unit during shutdown overrode the shutdown method. I agree that this seems counter-intuitive and unwanted. But I think the right solution here would be to ignore FailureAction= if a reboot is already in progress. I.e. instead of rejecting the job after it has been created, the creation of the job should be skipped.

Also, we already have ManagerObjective state, and this should be checked, instead of creating a new flag.

keszybz commented:
Oh, and please remove all the temporary parts like `####` and commented out fields — we never want to merge stuff like that.

ejb42 commented:
keszybz - Thank you for your review. I will work on your suggestion asap.

ejb42 commented:
Zbigniew,

I am trying to address your comments but I have run into two points of confusion.

I am thinking the manager_add_job() function is the point where the job gets created, is that not correct?
I thought the logic I added allowed the first shutdown target job to be created but not any other starting type of job.
Please point me to where the SPECIAL_(reboot/poweroff/halt)_TARGET jobs are created,  I am having trouble finding that.

I was not aware of the ManagerObjective, however it does not appear to change when the shutdown target is active.
In my testing, I added a change to the objective value once the shutdown target was created, but the manager_loop
exits and the shutdown does not happen like it should - none of the services get shutdown. Is that really something I should
use for this purpose?

Thanks,
-Jay
From: Zbigniew Jędrzejewski-Szmek <notifications@github.com>
Sent: Thursday, March 5, 2020 10:41 AM
To: systemd/systemd <systemd@noreply.github.com>
Cc: Burger, Jay <Jay.Burger@fujitsu.com>; Author <author@noreply.github.com>
Subject: Re: [systemd/systemd] honor-first-shutdown (#14945)


This solution seems pretty intrusive. The original problem was that a FailureAction= from a failing unit during shutdown overrode the shutdown method. I agree that this seems counter-intuitive and unwanted. But I think the right solution here would be to ignore FailureAction= if a reboot is already in progress. I.e. instead of rejecting the job after it has been created, the creation of the job should be skipped.

Also, we already have ManagerObjective state, and this should be checked, instead of creating a new flag.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/systemd/systemd/pull/14945?email_source=notifications&email_token=AOUM5OF7W3IEW6PFVSC6ZJ3RF7IYRA5CNFSM4K4IBPAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEN57NQQ#issuecomment-595326658>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOUM5ODHFHNAY4IKAB2PAF3RF7IYRANCNFSM4K4IBPAA>.

keszybz commented:
See `start_unit()` in systemctl.c. It just queues a halt.target/isolate job or similar.

ejb42 commented:
@keszybz  I believe I have addressed your rework concerns. I have tested the changes in the same fashion as documented. Please let me know if there are any other issues to address. Thanks.

keszybz commented on src/core/manager.h:
```diff
@@ -49,6 +49,7 @@ typedef enum ManagerObjective {
         MANAGER_HALT,
         MANAGER_KEXEC,
         MANAGER_SWITCH_ROOT,
+        MANAGER_SHUTDOWN,
```
"Shutdown" is either "halt" or "poweroff". I don't think there's any reason to add another enum here.

keszybz commented on src/core/unit.c:
```diff
@@ -2600,8 +2600,12 @@ void unit_notify(Unit *u, UnitActiveState os, UnitActiveState ns, UnitNotifyFlag
+                                emergency_action(m, u->failure_action, 0, u->reboot_arg, unit_failure_action_exit_status(u), reason);
+                        } else {
+                                log_info( "Tried to add Failure Action reboot job - Shutdown active" );
+                        }
```
Most of the other modifications in `src/core` are unexpected, and I think a localized change here is all that should be needed, or actually might be even nicer to modify emergency_action().

Basically, you want to have a condition like
```C
if (IN_SET(m->objective, MANAGER_REBOOT, MANAGER_POWEROFF, MANAGER_HALT, MANAGER_KEXEC))
    /* log a line that the emergency action is skipped because the manager is already terminating... */
else {
    /* do the old thing */
}
```

ejb42 commented:
@keszybz Thanks you for your reviews. I believe these changes are ready for review. 
Here is a summary:
- All of the changes for this are in emergency_action.c 
- I left a change to the order of the checks in manager_state() in manager.c.  It will report stopping before possibly reporting starting. I found a case where a failure which would cause a shutdown during the starting phase would cause manager_state() to return starting instead of stopping.
- Also fixed your comments on the test files.

keszybz commented on src/test/test-honor-first-shutdown.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+echo "Honor first shutdown test script"
+count=1
+while [ $count -eq 1 ]; do
```
What is the purpose of the $count variable? What is the purpose of the loop?

ejb42 replied:
Will remove the count. The loop provides a simulated main pid that can get killed by execstop to generate the FailureAction=reboot action during a poweroff initiated by the end.service in the added tests.

keszybz commented on src/test/test-honor-first-shutdown.service:
```diff
@@ -0,0 +1,11 @@
+After=multi-user.target
+
+[Service]
+ExecStart=/home/test-honor-first-shutdown.sh
```
`/home/` ?

ejb42 replied:
Changed to / per the added test

keszybz commented on src/test/test-honor-first-shutdown.service:
```diff
@@ -0,0 +1,11 @@
+FailureAction=reboot
+
+[Install]
+WantedBy=multi-user.target
```
This test is not hooked up anywhere that I can see... Is it supposed to be doing something?

ejb42 replied:
Created a new test/TEST-47-HONORFIRSTSHUTDOWN/ which will be seen when we resolve the question below.

keszybz commented on src/core/manager.c:
```diff
@@ -4245,6 +4245,11 @@ ManagerState manager_state(Manager *m) {
+        u = manager_get_unit(m, SPECIAL_SHUTDOWN_TARGET);
+        if (u && unit_active_or_pending(u))
+                return MANAGER_STOPPING;
+
```
Hmm, this change looks reasonable. But it needs to be a separate commit with an appropriate title and an explanation what issue it is solving and when that issue occurs.

ejb42 replied:
This has been done with PR #15330

keszybz commented on src/core/emergency-action.c:
```diff
@@ -27,11 +39,23 @@ void emergency_action(
 
+        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
+        u = manager_get_unit(m, SPECIAL_SHUTDOWN_TARGET);
+        if ((u && unit_active_or_pending(u))) {
```
So... why is manager_state() not used here?

ejb42 replied:
I think changing to manager_state would be the cleanest way as follows:
        /* Is the special shutdown target active or queued? If so, we are in shutdown state */
        if ((manager_state(m) == MANAGER_STOPPING) &&
            ((action == EMERGENCY_ACTION_REBOOT) ||
             (action == EMERGENCY_ACTION_POWEROFF))) {
                log_info("EmergencyAction: Shutdown is already active Skipping %s request",
                         emergency_action_table[action]);
                return;
        }
However, this change would not make it easy to use log_unit_info() per the below question.

keszybz commented on src/core/emergency-action.c:
```diff
@@ -27,11 +39,23 @@ void emergency_action(
+        if ((u && unit_active_or_pending(u))) {
+                if ( (action == EMERGENCY_ACTION_REBOOT) ||
+                     (action == EMERGENCY_ACTION_POWEROFF) ) {
+                        log_info("EmergencyAction Special Shutdown Tgt is active Skip %s",
```
Please no abbreviations like this. Use log_unit_info with a normal sentence.

ejb42 replied:
It does not look like emergency_action is built to use log_unit_info. If I remove the call to manager_get_unit() to use manager_state() there would be no unit available to call log_unit_info(). I can do as you request but I think I would still have to call manager_get_unit() in order to use log_unit_info(). Is that the way you would like to go with this?

keszybz commented:
> Delete manager.c 

Hmm, what is going on here?

ejb42 commented:
@keszybz 
I have rebased my fork to the latest systemd and as a result I have this branch in an un-recoverable state. I believe my only recourse at this point is to create a new branch and pull request which I will do named "honor-first-shutdown-request"

ejb42 commented:
Zbigniew,

I apologize for the confusion. That deletion was done out of my ignorance on how to
recover my mistake on that branch. I felt my best approach would be to create a new
branch and have opened PF #15400 to continue this change.
Please let me know if this is a problem and how I should proceed.

Thank you for your patience,
-Jay


From: Zbigniew Jędrzejewski-Szmek <notifications@github.com>
Sent: Friday, April 10, 2020 11:20 AM
To: systemd/systemd <systemd@noreply.github.com>
Cc: Burger, Jay <Jay.Burger@fujitsu.com>; Author <author@noreply.github.com>
Subject: Re: [systemd/systemd] honor-first-shutdown (#14945)


Delete manager.c

Hmm, what is going on here?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/systemd/systemd/pull/14945#issuecomment-612104565>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOUM5OFQHDDO4WC5SUDAFMTRL5BMLANCNFSM4K4IBPAA>.
